library(tidyverse)
#Import the data
datadir_wgcna <- "/home/gloria/Documentos/2.Master_Bioinfomatica_i_Bioestadistica/TFM/RNA_seq/counts_m/"

#list the files from the directory variable and convert to data frame that will be the phenodata
wgcna_phdata<-data.frame(fname=list.files(path=datadir_wgcna,pattern="*"),stringsAsFactors=FALSE)

#modify phenodata dataset to stablish for each sample the levels of the variables
wgcna_phdata <- wgcna_phdata %>% transmute(sample=substr(fname, 1,8),fname)
wgcna_phdata <- wgcna_phdata %>%
  mutate(map=as.factor(substr(fname,1,1)),
         trt=substr(fname,6,6),
         pop=substr(fname,3,4),
         md5=tools::md5sum(file.path(datadir_wgcna,fname)))



#1.NORMALIZATION =================================================================
library(DESeq2)
#to import count files generated by HTSeq.
wgcna_dds<-DESeqDataSetFromHTSeqCount(sampleTable=wgcna_phdata, directory=datadir_wgcna, design= ~1)

#Differential expressiona anlysis
wgcna_ddsDE <- DESeq(wgcna_dds)

#Extraction of log2foldChange and padj values
wgcna_ddsDE_results <- results(wgcna_ddsDE)


# Coerce to a data frame
wgcna_ddsDE_results_df <- as.data.frame(wgcna_ddsDE_results)

#Normalized counts
wgcna_counts_norm <- data.frame(counts(wgcna_ddsDE, normalized = TRUE))

##2.OUTLIERS DETECTION =============================================================
library(WGCNA)
#transpose the dataframe
t_wgcna_counts_norm <- data.frame(t(wgcna_counts_norm))

###outliers detection ---------------------------------------------------------

gsg <- goodSamplesGenes(t_wgcna_counts_norm)
summary(gsg)
gsg$allOK #True means no outliers
table(gsg$goodGenes)  
table(gsg$goodSamples)

datExpr = t_wgcna_counts_norm
nGenes = ncol(datExpr)
nSamples = nrow(datExpr)


#3.READ PHENOTYPIC DATA =============================================================
#IMport the data trait
traitData <- read.csv("/home/gloria/Documentos/1.PhD/4.rnaseq_SBG_GEO/WGCNA/traits_rnaseq_m.csv", sep = ",", header = TRUE)
Samples <- rownames(datExpr)
traitRows <- match(Samples, traitData$Deme)
datTraits <- traitData[traitRows, -1]
rownames(datTraits) <- traitData[traitRows, 1]
collectGarbage()

#To performed the sample dendogram and heatmap traits
sampleTree2 <- hclust(dist(datExpr), method = "average")
traitColors <- numbers2colors(datTraits, signed = FALSE)
plotDendroAndColors(sampleTree2, traitColors,
                    groupLabels = names(datTraits), 
                    main = "Sample dendrogram and trait heatmap")

##4.CONSTRUCT LA NETWORK ===============================================================
par(mfrow = c(2,1))
cex1 = 0.9
# Choose a set of soft-thresholding powers
powers = c(c(1:30), seq(from = 12, to=30, by=2))
#Call the network topology analysis function
sft = pickSoftThreshold (datExpr, powerVector = powers, verbose = 5)

sft_data <- sft$fitIndices

#visualization to pick power
plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit,signed R^2",type="n",
     main = paste("Scale independence"))
text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     labels=powers,cex=cex1,col="red")
abline(h=0.90,col="red") #This line corresponds to use a cut-off R² of h

plot(sft$fitIndices[,1], sft$fitIndices[,5],
     xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",
     main = paste("Mean connectivity"))
text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col="red")

#This line corresponds to use a cut-off R² of h
abline(h=0.90,col="red")

#Chose the soft power base on Scale independence plot and Mean connectivity plot
softPower = 6

#5.GROUPING GENES IN MODULES ------------------------------------------------------
#Construct the gene network
net <- blockwiseModules(datExpr, 
                        networkType = "signed",
                        minBlockSize = 20, 
                        TOMType = "unsigned", 
                        power = softPower,
                        mergeCutHeight = 0.25,
                        numericLabels = FALSE,
                        randomSeed = 10,
                        maxPOutliers = 0.01, 
                        verbose = 3) 
bwnet <- net

#get number of genes for each module
table(bwnet$colors)


#5.Module Eigengenes==============================================================
#4. Grouping Genes in Modules. 
module_eigengenes <- bwnet$MEs
head(module_eigengenes)

#Calculating the module dissimilarity eigengenes
MEDiss = 1-cor(module_eigengenes)

#Clustering the eigengenes modules
METree = hclust(as.dist(MEDiss), method = "average")

#Plotting the result
plot(METree, main = "Clustering of module eigengenes",
     xlab = "", sub = "")
#Grouping the clusters from a cut-off
MEDissThres = 0.01
#Plotting a cut-off line
abline(h=MEDissThres, col = "red")

#Grouping module colors
mergedColors = bwnet$colors

#Eigengenes of new grouped modules
mergedMEs = bwnet$MEs
bwLabels = matchLabels(bwnet$colors, moduleLabels)

moduleColors = mergedColors
colorOrder = c("grey", standardColors(50))
moduleLabels = match(moduleColors, colorOrder)-1
bwLabels = matchLabels(bwnet$colors, moduleLabels)
bwModuleColors = labels2colors(bwLabels)
bwModuleColors[bwnet$blockGenes[[1]]]

#Plot the final heatmap
plotDendroAndColors(bwnet$dendrograms, bwModuleColors[bwnet$blockGenes[[1]]],
                    "Module colors", main = "Gene dendrogram and module colors in #block 1", 
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05)







